### PR TITLE
uplink: change to firewall4

### PR DESCRIPTION
**Pull request by PolynomialDivision**

### CODE REVIEW

File: roles/cfg_openwrt/templates/uplink_gateway/nftables.d/22-samplicator-redirect.nft.j2 [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{% if openwrt_version == 'snapshot' %}
+chain samplicator_redirect {
+	type nat hook prerouting priority dstnat -1; policy accept;
+	oifname "lo" udp dport 3780 counter packets 0 bytes 0 redirect to :2000
```
Comment (spolack) [inline roles/cfg_openwrt/templates/uplink_gateway/nftables.d/22-samplicator-redirect.nft.j2:4]: why do you add the counter stats?

File: roles/cfg_openwrt/templates/uplink_gateway/nftables.d/21-limit-inbound.nft.j2 [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{% if openwrt_version == 'snapshot' %}
+chain limit_inbound {
+	type filter hook forward priority filter - 1; policy accept;
+	ct state established,related accept comment "Allow established connections"
+	tcp flags & ack == ack limit rate 5000/second burst 2500 packets counter packets 0 bytes 0 jump accept_to_freifunk comment "Allow established connections which not made it in the conntrack sync yet"
```
Comment (spolack) [inline roles/cfg_openwrt/templates/uplink_gateway/nftables.d/21-limit-inbound.nft.j2:5]: why do you add the counter stats?

File: roles/cfg_openwrt/templates/uplink_gateway/nftables.d/20-gre-maxseg-size.nft.j2 [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+{% set IPV4_HEADER_SIZE = 20 %}
+{% set IPV6_HEADER_SIZE = 40 %}
+{% set GRE_HEADER_SIZE = 24 %}
+{% set MTU = 1500 %}
+{% set ipv4_mss = MTU - GRE_HEADER_SIZE - TCP_HEADER_SIZE - IPV4_HEADER_SIZE %}
+{% set ipv6_mss = MTU - GRE_HEADER_SIZE - TCP_HEADER_SIZE - IPV6_HEADER_SIZE %}
+chain gre_maxseg {
+    type filter hook forward priority -2; policy accept;
+	oifname "gre*" tcp flags syn tcp option maxseg size set meta nfproto map { ipv4 : {{ ipv4_mss }}, ipv6 : {{ ipv6_mss }} }
+	iifname "gre*" tcp flags syn tcp option maxseg size set meta nfproto map { ipv4 : {{ ipv4_mss }}, ipv6 : {{ ipv6_mss }} }
```
Comment (spolack) [inline roles/cfg_openwrt/templates/uplink_gateway/nftables.d/20-gre-maxseg-size.nft.j2:12]: if you do oif on both sides, then iif is redundant

File: roles/cfg_openwrt/templates/uplink_gateway/nftables.d/20-gre-maxseg-size.nft.j2 [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+{% set TCP_HEADER_SIZE = 20 %}
+{% set IPV4_HEADER_SIZE = 20 %}
+{% set IPV6_HEADER_SIZE = 40 %}
+{% set GRE_HEADER_SIZE = 24 %}
+{% set MTU = 1500 %}
+{% set ipv4_mss = MTU - GRE_HEADER_SIZE - TCP_HEADER_SIZE - IPV4_HEADER_SIZE %}
+{% set ipv6_mss = MTU - GRE_HEADER_SIZE - TCP_HEADER_SIZE - IPV6_HEADER_SIZE %}
+chain gre_maxseg {
+    type filter hook forward priority -2; policy accept;
+	oifname "gre*" tcp flags syn tcp option maxseg size set meta nfproto map { ipv4 : {{ ipv4_mss }}, ipv6 : {{ ipv6_mss }} }
```
Comment (spolack) [inline roles/cfg_openwrt/templates/uplink_gateway/nftables.d/20-gre-maxseg-size.nft.j2:11]: why do you adjust ipv4 mss? we dont route over gre. sure it seems more complete but please write a comment its never being used

**Review by spolack — CHANGES_REQUESTED** [review]
some quick findings. didnt test though, and this has to be tested carefully

File: roles/cfg_openwrt/templates/uplink_gateway/nftables.d/20-gre-maxseg-size.nft.j2 (hunk 4 above)
Comment (PolynomialDivision) [reply roles/cfg_openwrt/templates/uplink_gateway/nftables.d/20-gre-maxseg-size.nft.j2:12]: I wrote a comment that is only visible in full file overview

**Comment by PolynomialDivision** [comment]
What about testin firewall4?

File: roles/cfg_openwrt/templates/uplink_gateway/nftables.d/20-gre-maxseg-size.nft.j2 [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{% if openwrt_version == 'snapshot' %}
+{% set TCP_HEADER_SIZE = 20 %}
+{% set IPV4_HEADER_SIZE = 20 %}
+{% set IPV6_HEADER_SIZE = 40 %}
+{% set GRE_HEADER_SIZE = 24 %}
+{% set MTU = 1500 %}
+{% set ipv4_mss = MTU - GRE_HEADER_SIZE - TCP_HEADER_SIZE - IPV4_HEADER_SIZE %}
+{% set ipv6_mss = MTU - GRE_HEADER_SIZE - TCP_HEADER_SIZE - IPV6_HEADER_SIZE %}
+chain gre_maxseg {
+    type filter hook forward priority -2; policy accept;
```
Comment (spolack) [inline roles/cfg_openwrt/templates/uplink_gateway/nftables.d/20-gre-maxseg-size.nft.j2:10]: why priority -2? It shouldnt matter at which position the mss is changed

File: roles/cfg_openwrt/templates/uplink_gateway/nftables.d/21-limit-inbound.nft.j2 [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{% if openwrt_version == 'snapshot' %}
+chain limit_inbound {
```
Comment (spolack) [inline roles/cfg_openwrt/templates/uplink_gateway/nftables.d/21-limit-inbound.nft.j2:2]: This whole file cannot work. Feels like this is a shortcoming of fw4. All hooking nftables chains are traversed until the packet is being rejected. A allow only allows passing the next chain until the final chain is reached, only then its truly allowed. Not sure how to fix that

Comment (spolack) [reply roles/cfg_openwrt/templates/uplink_gateway/nftables.d/21-limit-inbound.nft.j2:2]: someone already opened a ticket for that.
https://github.com/openwrt/openwrt/issues/9981

Comment (spolack) [reply roles/cfg_openwrt/templates/uplink_gateway/nftables.d/21-limit-inbound.nft.j2:2]: Feels we cannot switch to firewall4 as of now.

**Review by spolack — CHANGES_REQUESTED** [review]

**Pull request closed without merging**